### ISSUE TITLE
Error: GStreamer (unable to prepare device) [at Lubuntu]

### ISSUE THREAD

**Issue by AltS1** [nulloy/nulloy#65]
I don't really know how to reproduce it, happens from time to time. Sometimes after file switching, sometimes at the start, sometimes after using other players. "Log" window also doesn't help much because symbols there shows incorrectly

![7lxygwl](https://cloud.githubusercontent.com/assets/717109/8766244/755cb4cc-2e38-11e5-94cd-0bf128d807c2.png)

I can only guess that this error has something to do with .asoundrc file (using one that disables dmix http://www.slackwiki.com/Disable_dmix_for_alsa).

**Comment by noscript** [comment]
I haven't seen such ever. What is `/home/alt`? Some things to try:

* try to reproduce it with some other GStreamer based player, like Totem.
* run Nulloy with `GST_DEBUG=<level>` and try to find why the error happens. e.g. `GST_DEBUG=9 nulloy`. However level 9 is too verbose, so start with level 1 maybe and then go higher.

**Comment by AltS1** [comment]
'/home/alt' is my home folder under Lubuntu, It's noted probably because nulloy "shortcut" is placed at '/home/alt/.local/share/applications' (on image bellow there's '/usb/bin' from where I started player)

So here's something with GST_DEBUG=2:
![2015-07-20-104154_1440x900_scrot](https://cloud.githubusercontent.com/assets/13396238/8771604/5194831e-2ecc-11e5-9d9b-2d6093ccafad.png)

AL lib: (EE) alsa_reset_playback: snd_pcm_hw_params(data->pcmHandle, hp) failed: Файловий дескриптор у помилковому стані <translation: File descriptor in a wrong state>
0:00:40.263772914  4418 0xafc25780 WARN                  openal gstopenalsink.c:859:gst_openal_sink_prepare:<openalsink0> error: Unable to prepare device.
0:00:40.263858251  4418 0xafc25780 WARN                  openal gstopenalsink.c:859:gst_openal_sink_prepare:<openalsink0> error: ALC error: Invalid Device
0:00:40.264154165  4418 0xafc25780 WARN           audiobasesink gstaudiobasesink.c:1119:gst_audio_base_sink_preroll:<openalsink0> error: sink not negotiated.
0:00:40.265459894  4418 0xaf277d20 WARN               baseparse gstbaseparse.c:3264:gst_base_parse_loop:<mpegaudioparse13> error: streaming stopped, reason not-negotiated
0:00:40.340449180  4418  0x8f54a00 WARN                 basesrc gstbasesrc.c:3470:gst_base_src_start_complete:<source> pad not activated yet
....
0:00:40.919080904  4418 0xb082f860 WARN               baseparse gstbaseparse.c:3264:gst_base_parse_loop:<flacparse74> error: streaming stopped, reason not-negotiated
...
0:00:41.085492839  4418 0xaf266f80 WARN               baseparse gstbaseparse.c:3264:gst_base_parse_loop:<mpegaudioparse17> error: streaming stopped, reason not-negotiated
QFileInfo::absolutePath: Constructed with empty filename
...0:00:41.281662069  4418 0xaf266f80 WARN                   queue gstqueue.c:860:gst_queue_handle_sink_event:<aqueue> error: Помилка внутрішнього перенесення даних <translation: Error of internal data transfer>

0:00:41.281715853  4418 0xaf266f80 WARN                   queue gstqueue.c:860:gst_queue_handle_sink_event:<aqueue> error: streaming task paused, reason not-negotiated (-4)

**Comment by noscript** [comment]
I found similar reports, unrelated to GStreamer. That could be a bug in ALSA. What is your kernel version? Try to boot into some older kernel.

**Comment by noscript** [comment]
I pushed a2fc81a. Please try to build Nulloy locally and see if the log window now shows the symbols properly.

**Comment by AltS1** [comment]
I tried kernels 3.15, 3.19 and 4.1 - all the same. Interesting thing that even with the basic settings (http://www.alsa-project.org/main/index.php/Asoundrc#The_default_plugin) that doesn't disable dmix this error still occurs. Problem is pretty reproducible when you run the player right after system start.
 ---
Yes, symbols are fixed now. 
But there remains another (pretty common for both Windows and Linux) problem with tags where Cyrillic symbols are saved in Latin-1 (ISO-8859-1) codepage. In Audacious this problem could be fixed setting "WINDOWS-1251" codepage. Here's legal sample file http://www.horova.kiev.ua/song/blyz%27ko.mp3 to try.

**Comment by noscript** [comment]
I will add tags encoding selection, let's continue in a separate issue #67.

Could you please try reproduce the problem with Totem or Rhythmbox.

**Comment by AltS1** [comment]
Sadly Totem shows the same "unable to prepare device" when .asoundrc (which disables software mixing) is present.
But there's remain hang which happens when Nulloy tries to show "log" window. It has nothing to do with GStreamer so I'm unable to provide any useful log.

**Comment by noscript** [comment]
Any steps how to reproduce the freeze?

**Comment by AltS1** [comment]
Well, it's not that easy.
- Create .asoundrc file with content from above and make sure that you receive "unable to prepare device" warnings.
- Try to load at least 5 (more = bigger chance) of different files. If they'll play - clear play-list, restart player, load them again. Try few times until player would hang.
Or alternatively make some big playlist, close player, create .asoundrc, run player. If Nulloy wouldn't hang at first just double-click few times on some tracks in playlist.

**Comment by AltS1** [comment]
"Unable to prepare device" was fixed for me. Not sure what exactly fixed it but here's what I did:
- remove completely gstreamer (sudo apt-get purge gstreamer*)
- remove completely pulseaudio (sudo apt-get autoremove pulseaudio)
- installed gstreamer1.0-alsa (sudo apt-get install gstreamer1.0-alsa) When I removed PulseAudio sound didn't work without this plugin
Maybe the last step is enough (can't check because I've installed/uninstalled a lot of stuff) to fix the problem on vanilla Lubuntu. 
Also I noticed (using command "cat /proc/asound/card0/pcm0p/sub0/hw_params") that now output is more correct (ALSA-like). Previously it had fixed settings:
        format: S32_LE
        period_size 1024
        buffer_size 4096 
Now they're changing accordingly to file format, sample rate and bit-depth which is a good thing.

P.S. Noticed another easily-reproducible minor freeze with installed pulseaudio (sudo apt-get install pulseaudio).
Steps to reproduce:
- In Audacious choose ALSA output and play some file
- Open Nulloy and try to play some file. Nulloy UI would freeze until you release device from Audacious (pressing stop or closing player).

**Comment by noscript** [comment]
I tested with a fresh Lubutnu 15.04 installation but I couldn't reproduce any of the described problems. Adding or removing `pulseaudio`, editing `~/.asoundrc` or using Audacious with ALSA or Pulseaudio output in parallel with Nulloy - everything works fine for me.

I suggest you to try to reproduce it with a clean Lubutnu installation too (in Virtualbox maybe).

**Comment by AltS1** [comment]
Re-produced initial error (unable to prepare device) and hang with LiveCD of Lubuntu 15.04. 
Also tried combination of PulseAudio and Audacity (ALSA) and nothing wrong happened.

**Comment by noscript** [comment]
> Re-produced initial bug (unable to prepare device and further hang) with LiveCD of Lubuntu 15.04. 

Was in a virtual machine? The point is to try with emulated soundcard. Otherwise it could be that your real sound card has some issues with ALSA.

Here is my step-by-step:

1 . Fresh Lubuntu x64 install in Virtualbox (I don't have too much RAM for doing it in LiveCD).
2 . Nulloy 0.8 install from repository.
3 . `sudo apt-get autoremove pulseaudio` reports `Package 'pulseaudio' is not installed, so not removed`
4 . `sudo apt-get install gstreamer1.0-alsa`
5 . Created `~/.asoundrc` with:
```
pcm.!default {
	type hw
	card 0
}
ctl.!default {
	type hw           
	card 0
}
```
6 . Rebooted
7 . Downloaded a bunch of mp3's from soundcloud and played in Nulloy. No problem.

Continuing with Audacity:
7 . `sudo apt-get install pulseaudio audacity`
8 . Rebooted
9 . Running Audacity, checked that ALSA is set for output in Preferences.
10 . Playing files simultaneously in Audacity and Nulloy. No problem.

When Nulloy complains with "unable to prepare device", try `gst-launch-1.0`:
```
gst-launch-1.0 playbin uri=file:///some/file.mp3 
```

**Comment by AltS1** [comment]
Sorry for the late answer. 
Yes, it may be problem with ALSA, GStreamer or Lubuntu (I tried x86 btw) but if I can fix it that really doesn't matter (also it seems like initial problem is not player's fault). 
What really matters is that Nulloy hangs when something goes wrong. I remember one time it even hanged when there were missing files in playlist (again, it's not reproducible). So I propose ability to write debug log (for player only) so users can show you what exactly happens.

**Comment by noscript** [comment]
>What really matters is that Nulloy hangs when something goes wrong. I remember one time it even hanged when there were missing files in playlist (again, it's not reproducible). So I propose ability to write debug log (for player only) so users can show you what exactly happens.

It is extremely rare that I cannot reproduce the reported problem. Even if that really happens, a simple debug log would help only little.

I agree, it would be great to have an automatically generated report e.g. when player crashes, something like what macOS does. But I cannot come up with a cross-platform solution. I'm not sure how to approach this.

Either way, please try the latest release. Hopefully it works better.
